iocuddle/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+    ///
+    /// # Safety
+    ///
+    /// For safety details, see [Ioctl::classic].
+    pub const unsafe fn write_read<'a, T>(self, nr: u8) -> Ioctl<WriteRead, &'a T> {
+        self.make(nr, 0b11)
+    }
+}
+
+/// A defined `ioctl` along with its associated `direction` and `type`
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct Ioctl<D, T>(c_ulong, PhantomData<(D, T)>);
+
+impl<D, T> Ioctl<D, T> {
+    /// Define a new `ioctl` with an associated `direction` and `type`
+    ///
+    /// The `request` argument is the allocated integer which uniquely
+    /// identifies this `ioctl`.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because defining an `ioctl` with an incorrect
+    /// `request`, `direction` or argument `type` can later result in memory
+    /// corruption. You are responsible to ensure these values are correct.
+    ///
+    /// Further, you are responsible to ensure that the argument `type` itself
+    /// provides appropriate safe wrappers around its raw contents. For some
+    /// `type`s none are required. For others, particularly `type`s that pass
+    /// pointers to the kernel as `u64`, you need to ensure that things like
+    /// lifetimes are correct.
+    pub const unsafe fn classic(request: c_ulong) -> Self {
+        Self(request, PhantomData)
+    }
+}
+
+impl Ioctl<Read, c_void> {
+    /// Issue an `ioctl` to read a file descriptor's metadata as `c_uint`.
+    ///
+    /// No argument is supplied to the internal `ioctl()` call. The raw
+    /// (positive) return value from the internal `ioctl()` call is returned
+    /// on success.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl<T> Ioctl<Read, &T> {
+    /// Issue an `ioctl` to read a file descriptor's metadata as type `T`.
+    ///
+    /// A zeroed instance of type `T` is passed as the first argument to the
+    /// internal `ioctl()` call. Upon success, returns the raw (positive)
+    /// return value and the instance of `T`.
+    pub fn ioctl(self, fd: &impl AsRawFd) -> Result<(c_uint, T)> {
+        let mut out: T = unsafe { core::mem::MaybeUninit::zeroed().assume_init() };
+
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, &mut out as *mut _, null::<c_void>()) };
+
+        r.try_into()
+            .and_then(|x| Ok((x, out)))
+            .or_else(|_| Err(Error::last_os_error()))
+    }
+}
+
+impl Ioctl<Write, c_void> {
+    /// Issue an `ioctl` to modify a file descriptor
+    ///
+    /// No argument is provided.
+    ///
+    /// On success, returns the (positive) return value.
+    pub fn ioctl(self, fd: &mut impl AsRawFd) -> Result<c_uint> {
+        let r = unsafe { ioctl(fd.as_raw_fd(), self.0, null::<c_void>()) };
+
+        r.try_into().or_else(|_| Err(Error::last_os_error()))
```

Comment:
Could be simplified to `.map_err(|_| Error::last_os_error())`

Reply:
Fixed.